x/stake/keeper/validator.go

```diff
@@ -84,23 +98,109 @@ func (k Keeper) SetValidatorByConsAddr(ctx sdk.Context, validator types.Validato
-func (k Keeper) validatorByPowerIndexExists(ctx sdk.Context, power []byte) bool {
+// validator index
+func (k Keeper) SetNewValidatorByPowerIndex(ctx sdk.Context, validator types.Validator) {
 	store := ctx.KVStore(k.storeKey)
-	return store.Get(power) != nil
+	pool := k.GetPool(ctx)
+	store.Set(GetValidatorsByPowerIndexKey(validator, pool), validator.OperatorAddr)
+}
+
+//___________________________________________________________________________
+
+// Update the tokens of an existing validator, update the validators power index key
+func (k Keeper) AddValidatorTokensAndShares(ctx sdk.Context, validator types.Validator,
+	tokensToAdd sdk.Int) (valOut types.Validator, addedShares sdk.Dec) {
+
+	pool := k.GetPool(ctx)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+	validator, pool, addedShares = validator.AddTokensFromDel(pool, tokensToAdd)
+	// increment the intra-tx counter
+	// in case of a conflict, the validator which least recently changed power takes precedence
+	counter := k.GetIntraTxCounter(ctx)
+	validator.BondIntraTxCounter = counter
+	k.SetIntraTxCounter(ctx, counter+1)
+	k.SetValidator(ctx, validator)
+	k.SetPool(ctx, pool)
+	k.SetValidatorByPowerIndex(ctx, validator, pool)
+	return validator, addedShares
+}
+
+// Update the tokens of an existing validator, update the validators power index key
+func (k Keeper) RemoveValidatorTokensAndShares(ctx sdk.Context, validator types.Validator,
+	sharesToRemove sdk.Dec) (valOut types.Validator, removedTokens sdk.Dec) {
+
+	pool := k.GetPool(ctx)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+	validator, pool, removedTokens = validator.RemoveDelShares(pool, sharesToRemove)
+	k.SetValidator(ctx, validator)
+	k.SetPool(ctx, pool)
+	k.SetValidatorByPowerIndex(ctx, validator, pool)
+	return validator, removedTokens
+}
+
+// Update the tokens of an existing validator, update the validators power index key
+func (k Keeper) RemoveValidatorTokens(ctx sdk.Context, validator types.Validator, tokensToRemove sdk.Dec) types.Validator {
+	pool := k.GetPool(ctx)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
+	validator, pool = validator.RemoveTokens(pool, tokensToRemove)
+	k.SetValidator(ctx, validator)
+	k.SetPool(ctx, pool)
+	k.SetValidatorByPowerIndex(ctx, validator, pool)
+	return validator
+}
+
+// UpdateValidatorCommission attempts to update a validator's commission rate.
+// An error is returned if the new commission rate is invalid.
+func (k Keeper) UpdateValidatorCommission(ctx sdk.Context, validator types.Validator, newRate sdk.Dec) (types.Commission, sdk.Error) {
+	commission := validator.Commission
+	blockTime := ctx.BlockHeader().Time
+
+	if err := commission.ValidateNewRate(newRate, blockTime); err != nil {
+		return commission, err
+	}
+
+	commission.Rate = newRate
+	commission.UpdateTime = blockTime
+
+	return commission, nil
+}
+
+// remove the validator record and associated indexes
+func (k Keeper) RemoveValidator(ctx sdk.Context, address sdk.ValAddress) {
```

Comment:
we should probably note that this doesn't remove the `BondedValidatorIndex`- but arguably it should - we could pass in a bool to remove this index in this function? thoughts? I'll leave it to you, maybe it's fine as it - but it would be nice to have all the indices removed in a consolidated location.

Reply:
Hmm, I like only handling the `BondedValidatorIndex` in `EndBlock`, I think it's simpler - presently we don't need to deal with that index anywhere else.

Reply:
cool well we should at least add a note on the comment which says `and associated indexes` - which is not totally true if we leave the one index out